### ISSUE TITLE
Best practices (initial dump)

### ISSUE THREAD

**Issue by alexandru-dinu** [opskrift/opskrift#27]
## Goals
- reusable code -- packages
- guidelines and workflows
- **programming with guarantees**
- https://github.com/graviraja/MLOps-Basics

## Experiment definition
- An experiment must be completely defined by hyperparameters, and uniquely identified by an ID.
- **Input:** hyperparameters
- **Output:** artifacts (trained model, metrics etc.)

## Processing pipelines
- Processing should be viewed as a pipeline, each stage being fully defined by the stages behind it.
- Each stage component expects and input and guarantees an output (modularity).
- This ensures that one component can be changed without breaking other components.
- Starting a new project $\to$ implementing pipeline stages / components.

## Saving & Loading
- Loading an experiment object should correctly initialize / load all related variables / datasets
- save `state_dict` without model class
- define models in a separate dir
- refer to https://stackoverflow.com/a/66652939
- For cases when the `forward` method is fixed, i.e. only the model arch is changed, then we can do this: https://stackoverflow.com/a/66954874

## Querying
- All assets should be linked to an experiment
- **Cleaning after failed experiments**
- suppose you start an experiment, but it won't finish gracefully (e.g. due to some Python error)
- you don't want to manually clean this experiment's data
- what's the best way to approach this?

## File management
1. save (large binary) files together with metadata describing how they were generated
    - must be easily readable without touching the actual data
2. naming scheme for multiple variants of the same conceptual data (maybe an UUID?)
    - primarily avoids putting args in filename, e.g. `dataset_user_splits_2020-2021_neg_factor_4`
3. (for sequential data): fragmented load, e.g. `load(obj + offset, obj + offset + size)`
4. have some sort of checkpoint (or similar) when processing and dumping large data
    - the idea is to have a fallback for cases when something fails after a lot of processing
    - can we abstract this?

## Serving
- See [pytorch/serve](https://github.com/pytorch/serve)

## Good practices
- Seed everything
- Checkpoints
- Hyper-parameter tuning
- Painless cross-validation
- Use pure python as much as possible (i.e. minimize dependencies).
- `argparse.Namespace`
    -  Managing paths in (large) projects
- esp. relative paths that are crucial for certain objects (e.g. `wppl` files in `pplib`)
    - Easy to visualize results
    - e.g. don't hardcode details into filenames

## PyTorch datamodel
```
(data sources) -> Dataset -> (__getitem__) -> (sampler / collate_fn) -> DataLoader -> (batch)
```
